core/telemetry/src/lib.rs

```diff
@@ -43,66 +48,165 @@ pub type Telemetry = slog_scope::GlobalLoggerGuard;
 
+/// Log levels.
+pub const SUBSTRATE_DEBUG: &str = "9";
+pub const SUBSTRATE_INFO: &str = "0";
+
+pub const CONSENSUS_TRACE: &str = "9";
+pub const CONSENSUS_DEBUG: &str = "5";
+pub const CONSENSUS_WARN: &str = "4";
+pub const CONSENSUS_INFO: &str = "3";
+
+/// Multiply logging to all drains. This is similar to `slog::Duplicate`, which is
+/// limited to two drains though and doesn't support dynamic nesting at runtime.
+#[derive(Debug, Clone)]
+pub struct Multiply<D: Drain> (pub Vec<Box<D>>);
+
+impl<D: Drain> Multiply<D> {
+	pub fn new(v: Vec<Box<D>>) -> Self {
+		Multiply(v)
+	}
+}
+
+impl<D: Drain> Drain for Multiply<D> {
+	type Ok = Vec<D::Ok>;
+	type Err = Vec<D::Err>;
+
+	fn log(&self, record: &Record, logger_values: &OwnedKVList) -> result::Result<Self::Ok, Self::Err> {
+		let mut oks = Vec::new();
+		let mut errs = Vec::new();
+
+		self.0.iter().for_each(|l| {
+			let res: Result<<D as Drain>::Ok, <D as Drain>::Err> = (*l).log(record, logger_values);
+			match res {
+				Ok(o) => oks.push(o),
+				Err(e) => errs.push(e),
+			}
+		});
+
+		if !errs.is_empty() {
+			result::Result::Err(errs)
+		} else {
+			result::Result::Ok(oks)
+		}
+	}
+}
+
 /// Initialise telemetry.
 pub fn init_telemetry(config: TelemetryConfig) -> slog_scope::GlobalLoggerGuard {
-	let writer = TelemetryWriter::new();
-	let out_sync = writer.out.clone();
-	let log = slog::Logger::root(
-		slog_async::Async::new(
-			slog_json::Json::default(writer).fuse()
-		).chan_size(CHANNEL_SIZE)
+	let mut endpoint_drains: Vec<Box<slog::Filter<_, _>>> = Vec::new();
+	let mut out_syncs = Vec::new();
+
+	// Set up a filter/drain for each endpoint
+	config.endpoints.0.iter().for_each(|(url, verbosity)| {
+		let writer = TelemetryWriter::new(Arc::new(url.to_owned()));
+		let out_sync = writer.out.clone();
+		out_syncs.push(out_sync);
+
+		let until_verbosity = *verbosity;
+		let filter = slog::Filter(
+			slog_json::Json::default(writer).fuse(),
+			move |rec| {
+				let tag = rec.tag().parse::<u8>()
+					.expect("`telemetry!` macro requires tag.");
+				tag <= until_verbosity
+			});
+
+		let filter = Box::new(filter) as Box<slog::Filter<_, _>>;
+		endpoint_drains.push(filter);
+	});
+
+	// Set up logging to all endpoints
+	let drain = slog_async::Async::new(Multiply::new(endpoint_drains).fuse());
+	let root = slog::Logger::root(drain.chan_size(CHANNEL_SIZE)
 		.overflow_strategy(slog_async::OverflowStrategy::DropAndReport)
 		.build().fuse(), o!()
 	);
-	let logger_guard = slog_scope::set_global_logger(log);
-
-	thread::spawn(move || {
-		loop {
-			trace!(target: "telemetry", "Connecting to Telemetry... {:?}", config.url);
-			let _ = ws::connect(config.url.as_str(), |out| Connection::new(out, &*out_sync, &config));
-
-			thread::sleep(time::Duration::from_millis(5000));
-		}
+	let logger_guard = slog_scope::set_global_logger(root);
+
+	// Spawn a thread for each endpoint
+	let on_connect = Arc::new(config.on_connect);
+	config.endpoints.0.iter().for_each(|(url, verbosity)| {
```

Comment:
Regarding the `Arc<String>` stuff you need to do the following:
```suggestion
	config.endpoints.0.into_iter().for_each(|(url, verbosity)| {
```

Then remove all this clones, `Arc::new`, and `Arc::clone`.